android/src/test/java/com/expensify/livemarkdown/RangeSplitterTest.java

```diff
@@ -0,0 +1,89 @@
+package com.expensify.livemarkdown;
+
+import static com.expensify.livemarkdown.RangeSplitter.splitRangesOnEmojis;
+
+
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class RangeSplitterTest {
+
+  private void testRange(MarkdownRange range, int start, int end, String type) {
+    assertEquals(start, range.getStart());
+    assertEquals(end, range.getEnd());
+    assertEquals(type, range.getType());
+  }
+
+  @Test
+  public void testNoOverlap() {
+    List<MarkdownRange> markdownRanges = new ArrayList<>();
+    markdownRanges.add(new MarkdownRange("strikethrough", 0, 10, 1));
+    markdownRanges.add(new MarkdownRange("emoji", 12, 2, 1));
+
+    splitRangesOnEmojis(markdownRanges, "strikethrough");
+
+    assertEquals(2, markdownRanges.size());
+
+    testRange(markdownRanges.get(0), 0, 10, "strikethrough");
+    testRange(markdownRanges.get(1), 12, 14, "emoji");
+  }
+
+  @Test
+  public void testOverlapWrongType() {
+    List<MarkdownRange> markdownRanges = new ArrayList<>();
+    markdownRanges.add(new MarkdownRange("strikethrough", 0, 10, 1));
+    markdownRanges.add(new MarkdownRange("emoji", 3, 4, 1));
+
+    splitRangesOnEmojis(markdownRanges, "italic");
+
+    assertEquals(2, markdownRanges.size());
+    testRange(markdownRanges.get(0), 0, 10, "strikethrough");
+    testRange(markdownRanges.get(1), 3, 7, "emoji");
+  }
+
+  @Test
+  public void testSingleOverlap() {
+    List<MarkdownRange> markdownRanges = new ArrayList<>();
+    markdownRanges.add(new MarkdownRange("strikethrough", 0, 10, 1));
+    markdownRanges.add(new MarkdownRange("emoji", 3, 4, 1)); // This should split the strikethrough range
+
+    splitRangesOnEmojis(markdownRanges, "strikethrough");
+
+    // Sort is needed because ranges may get mixed while splitting
+    Collections.sort(markdownRanges, (r1, r2) -> Integer.compare(r1.getStart(), r2.getStart()));
```

Comment:
Is it sufficient to compare ranges only based on `start`?

Is `splitRangesOnEmojis` deterministic?

Do we really need to sort the ranges?

Reply:
> Is it sufficient to compare ranges only based on start?

For the tests I wrote - I think so but in generality no.

> Is splitRangesOnEmojis deterministic?

It is.

> Do we really need to sort the ranges?

`splitRangesOnEmojis` does not guarantee order of ranges. I've added sort so that tests work for different implementation of the method.